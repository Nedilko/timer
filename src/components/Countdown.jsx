import { useState, useEffect, useContext } from "react";
import { calcDateDuration } from "../utils/dateUtils";
import {
  addMinutesSeconds,
  formatNumberTwoDigit,
} from "../utils/helperFunctions";
import PropTypes from "prop-types";

Countdown.propTypes = {
  targetDate: PropTypes.object.isRequired,
};

function Countdown({ targetDate }) {
  const fullDate = addMinutesSeconds(targetDate);
  const [timeLeft, setTimeLeft] = useState(calcDateDuration(fullDate));

  useEffect(() => {
    const timerId = setTimeout(() => {
      setTimeLeft((oldState) => calcDateDuration(fullDate));
    }, 1000);

    return () => {
      clearTimeout(timerId);
    };
  }, [timeLeft, targetDate]);

  return (
    <div className="flex-row flex-wrap items-end justify-around text-center uppercase">
      {timeLeft.days > 0 && (
        <div className="after:absolute after:text-2xl after:content-[':'] after:absolute after:right-0 after:bottom-0 relative inline-flex flex-col w-1/4 px-1">
          <div>days</div>
          <div className="text-2xl">{formatNumberTwoDigit(timeLeft.days)}</div>
        </div>
      )}

      {(timeLeft.hours > 0 || timeLeft.days > 0) && (
        <div className="after:absolute after:text-2xl after:content-[':'] after:absolute after:right-0 after:bottom-0 relative inline-flex flex-col w-1/4 px-1">
          <div>hours</div>
          <div className="text-2xl">{formatNumberTwoDigit(timeLeft.hours)}</div>
        </div>
      )}

      {(timeLeft.hours > 0 || timeLeft.days > 0 || timeLeft.minutes > 0) && (
        <div className="after:absolute after:text-2xl after:content-[':'] after:absolute after:right-0 after:bottom-0 relative inline-flex flex-col w-1/4 px-1">
          <div>minutes</div>
          <div className="text-2xl">
            {formatNumberTwoDigit(timeLeft.minutes)}
          </div>
        </div>
      )}

      <div className="inline-flex flex-col w-1/4 px-1">
        <div>seconds</div>
        <div className="text-2xl">{formatNumberTwoDigit(timeLeft.seconds)}</div>
      </div>
    </div>
  );
}

export default Countdown;
