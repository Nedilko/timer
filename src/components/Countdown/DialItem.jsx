import PropTypes from "prop-types";
import { formatNumberTwoDigit } from "../../utils/helperFunctions";

DialItem.propTypes = {
  title: PropTypes.string.isRequired,
  numbers: PropTypes.number.isRequired,
  isLast: PropTypes.bool,
};

function DialItem({ title, numbers, isLast }) {
  return (
    <div
      className={`inline-flex flex-col w-1/4 px-1 ${
        isLast
          ? ""
          : "after:absolute after:text-2xl after:content-[':'] after:absolute after:right-0 after:bottom-0 relative"
      }`}
    >
      <div>{title}</div>
      <div className="text-2xl">{formatNumberTwoDigit(numbers)}</div>
    </div>
  );
}

export default DialItem;
