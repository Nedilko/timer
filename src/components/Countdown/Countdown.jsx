import { useState, useEffect, useContext } from "react";
import { calcDateDuration } from "../../utils/dateUtils";
import {
  addMinutesSeconds,
  formatNumberTwoDigit,
} from "../../utils/helperFunctions";
import PropTypes from "prop-types";
import DialItem from "./DialItem";

Countdown.propTypes = {
  targetDate: PropTypes.object.isRequired,
  handleFinish: PropTypes.func.isRequired
};

function Countdown({ targetDate, handleFinish }) {
  const fullDate = addMinutesSeconds(targetDate);
  const [timeLeft, setTimeLeft] = useState(calcDateDuration(fullDate));

  useEffect(() => {
    const timerId = setInterval(() => {
      setTimeLeft((oldState) => calcDateDuration(fullDate));

      if ( timeLeft.days == 0 && timeLeft.hours == 0 && timeLeft.minutes == 0 && timeLeft.seconds == 0 ) {
        handleFinish(true)
      }

    }, 1000);

    return () => {
      clearInterval(timerId);
    };
  }, [timeLeft, targetDate]);

  return (
    <div className="flex-row flex-wrap items-end justify-around text-center uppercase">
      {timeLeft.days > 0 && (
        <DialItem title="days" numbers={timeLeft.days} isLast={false}/>
      )}

      {(timeLeft.hours > 0 || timeLeft.days > 0) && (
        <DialItem title="hours" numbers={timeLeft.hours} isLast={false}/>
      )}

      {(timeLeft.hours > 0 || timeLeft.days > 0 || timeLeft.minutes > 0) && (
        <DialItem title="minutes" numbers={timeLeft.minutes} isLast={false}/>
      )}

      <DialItem title="seconds  " numbers={timeLeft.seconds} isLast={true}/>
    </div>
  );
}

export default Countdown;
