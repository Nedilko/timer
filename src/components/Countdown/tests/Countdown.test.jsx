import { render, screen } from "@testing-library/react";
import Countdown from "../Countdown";

describe("Countdown", () => {
const handleFinish = jest.fn();
const targetDate = {
    days: 0,
    hours: 0,
    minutes: 0,
    seconds: 0
}

beforeEach(()=>{
    handleFinish.mockClear()
})

  it("should match snapshot", () => {
    render(<Countdown targetDate={targetDate} handleFinish={handleFinish}/>)

    expect(screen.getByRole("countdown")).toMatchSnapshot()
  });
  it("should handleFinish", () => {
    render(<Countdown targetDate={targetDate} handleFinish={handleFinish}/>)



  });
  it("targetDate", () => {});
});