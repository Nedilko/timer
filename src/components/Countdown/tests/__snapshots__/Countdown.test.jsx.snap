// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Countdown should match snapshot 1`] = `
<div
  class="flex-row flex-wrap items-end justify-around text-center uppercase"
  role="countdown"
>
  <div
    class="inline-flex flex-col w-1/4 px-1 after:absolute after:text-2xl after:content-[':'] after:absolute after:right-0 after:bottom-0 relative"
    role="dialitem"
  >
    <div>
      hours
    </div>
    <div
      class="text-2xl"
    >
      14
    </div>
  </div>
  <div
    class="inline-flex flex-col w-1/4 px-1 after:absolute after:text-2xl after:content-[':'] after:absolute after:right-0 after:bottom-0 relative"
    role="dialitem"
  >
    <div>
      minutes
    </div>
    <div
      class="text-2xl"
    >
      53
    </div>
  </div>
  <div
    class="inline-flex flex-col w-1/4 px-1 "
    role="dialitem"
  >
    <div>
      seconds  
    </div>
    <div
      class="text-2xl"
    >
      02
    </div>
  </div>
</div>
`;
